go/vt/vtgate/executor.go

```diff
@@ -1502,6 +1497,88 @@ func buildVarCharRow(values ...string) []sqltypes.Value {
 	return row
 }
 
+func generateCharsetRows(showFilter *sqlparser.ShowFilter, colNames []string) [][]sqltypes.Value {
+	if showFilter == nil {
+		return buildCharsetRows("both")
+	}
+
+	var filteredColName string
+	if showFilter.Like != "" {
+		filteredColName = checkLikeOpt(showFilter.Like, colNames)
+	}
+
+	cmpExp, ok := showFilter.Filter.(*sqlparser.ComparisonExpr)
+	if ok {
+		left, ok := cmpExp.Left.(*sqlparser.ColName)
+		if !ok {
+			panic("expect left side to be 'charset'")
+		}
+		leftOk := left.Name.EqualString("charset")
+
+		if leftOk {
+			sqlVal, ok := cmpExp.Right.(*sqlparser.SQLVal)
+			if !ok {
+				panic("we expect the right side to be a string")
+			}
+			rightString := string(sqlVal.Val)
+
+			switch cmpExp.Operator {
+			case "=":
+				for _, colName := range colNames {
+					if rightString == colName {
+						filteredColName = colName
+					}
+				}
+			case "like":
+				filteredColName = checkLikeOpt(rightString, colNames)
+			}
+		}
+
+	}
+	
+	return buildCharsetRows(filteredColName)
+}
+
+func buildCharsetRows(colName string) [][]sqltypes.Value {
+	rows := make([][]sqltypes.Value, 0, 4)
+	row0 := buildVarCharRow(
+		"utf8",
+		"UTF-8 Unicode",
+		"utf8_general_ci")
+	row0 = append(row0, sqltypes.NewInt32(3))
+	row1 := buildVarCharRow(
+		"utf8mb4",
+		"UTF-8 Unicode",
+		"utf8mb4_general_ci")
+	row1 = append(row1, sqltypes.NewInt32(4))
+
+	switch colName {
+	case "utf8":
+		rows = append(rows, row0)
```

Comment:
nit: instead of creating the rows and then appending to it, why not simply return a literal array?

```
	case "utf8":
		return [][]sqltypes.Value{row0}
```